internal/components/datasource/schema.go

```diff
@@ -2,124 +2,32 @@ package datasource
 
 import (
 	"embed"
-	"reflect"
-
-	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
+	"path/filepath"
 
+	"github.com/grafana/grafana/internal/cuectx"
 	"github.com/grafana/thema"
 )
 
-const (
-	cuePath     = "internal/components/datasource"
-	cueFilename = "lineage.cue"
-	schemaName  = "datasource"
-	groupName   = "datasource.core.grafana"
-	// TODO this should be derived from the Thema synv.
-	groupVersion = "v1alpha1"
-)
-
 var (
 	//go:embed lineage.cue
 	cueFS embed.FS
 
 	// TODO: this should be generated by Thema.
-	schemaVersion = thema.SV(0, 0)
-
-	// This schema is used for registering CRDs.
-	//
-	// TODO: this should be generated by Thema.
-	schemaOpenapi = apiextensionsv1.JSONSchemaProps{
-		Type: "object",
-		Properties: map[string]apiextensionsv1.JSONSchemaProps{
-			"spec": {
-				Type: "object",
-				Properties: map[string]apiextensionsv1.JSONSchemaProps{
-					"type": {
-						Type: "string",
-					},
-					"typeLogoUrl": {
-						Type: "string",
-					},
-					"access": {
-						Type: "string",
-					},
-					"url": {
-						Type: "string",
-					},
-					"password": {
-						Type: "string",
-					},
-					"user": {
-						Type: "string",
-					},
-					"database": {
-						Type: "string",
-					},
-					"basicAuth": {
-						Type: "boolean",
-					},
-					"basicAuthUser": {
-						Type: "string",
-					},
-					"basicAuthPassword": {
-						Type: "string",
-					},
-					"withCredentials": {
-						Type: "boolean",
-					},
-					"isDefault": {
-						Type: "boolean",
-					},
-					"jsonData": {
-						Type: "string",
-					},
-					"version": {
-						Type: "integer",
-					},
-					"readOnly": {
-						Type: "boolean",
-					},
-				},
-			},
-		},
-	}
+	currentVersion = thema.SV(0, 0)
 )
 
-// Datasource is the Kubernetes-compatible Datasource definition.
-//
-// TODO: this should be generated by Thema.
-type Datasource struct {
-	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata,omitempty"`
-
-	Spec   DatasourceSpec   `json:"spec,omitempty"`
-	Status DatasourceStatus `json:"status,omitempty"`
-}
-
-// DeepCopyObject returns a deep copy of Datasource.
-//
-// TODO: this should be generated by Thema.
-func (in *Datasource) DeepCopyObject() runtime.Object {
-	val := reflect.ValueOf(in).Elem()
-
-	cpy := reflect.New(val.Type())
-	cpy.Elem().Set(val)
-
-	ret, ok := cpy.Interface().(runtime.Object)
-	if !ok {
-		return nil
-	}
-
-	return ret
+// Lineage returns the Thema lineage representing Grafana datasources. The
+// lineage is the canonical specification of the datasource schema, all past
+// schema versions, and mappings that allow migration between schema versions.
+func Lineage(lib thema.Library, opts ...thema.BindOption) (thema.Lineage, error) {
+	return cuectx.LoadGrafanaInstancesWithThema(filepath.Join("internal", "components", "datasource"), cueFS, lib, opts...)
```

Comment:
Can this be extracted into a loader instead?

Reply:
Isn't it a loader already?

Reply:
Oh wait, i see what you're saying - no, i think this is probably preferable as-is, because it's following Thema's `LineageFactory` [pattern](https://github.com/grafana/thema/blob/main/docs/go-mapping.md#goal-the-lineage-factory) - the canonical way of exposing a Thema lineage in Go.

Exposing this will eventually be important for a [public] audience that wants to directly rely on the lineage.

Reply:
Example: when i write a Go library for generating dashboards, it'll never be used by the Go backend (and therefore needs to be ergonomic/independent of wire). Instead, i'll want to call this `Lineage()` func directly.

Reply:
Okay, that's fair. I would like to avoid as much boilerplate as possible, but I think this particular case should be fixed by codegen instead.